File: backend/src/core/audit.ts
import { appendFile } from "node:fs/promises";
import path from "node:path";
import { randomUUID } from "node:crypto";
import { env } from "../lib/env.ts";
import { ensureDir } from "../lib/fs.ts";

const sensitiveKeys = new Set(["authorization", "token", "password", "api_key", "anthropic_api_key"]);

function redact(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(redact);
  }

  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).map(([key, nested]) => [
        key,
        sensitiveKeys.has(key.toLowerCase()) ? "<redacted>" : redact(nested)
      ])
    );
  }

  return value;
}

export async function audit(event: string, payload: Record<string, unknown>) {
  const requestId = typeof payload.request_id === "string" ? payload.request_id : randomUUID();
  const entry = {
    ts: new Date().toISOString(),
    event,
    request_id: requestId,
    payload: redact(payload)
  };
  const auditDir = path.join(env.logDir, "audit");
  await ensureDir(auditDir);
  await appendFile(path.join(auditDir, "audit.jsonl"), `${JSON.stringify(entry)}\n`, "utf8");
  return requestId;
}

File: backend/app/core/audit.py
from datetime import datetime, timezone
from pathlib import Path
import json, uuid
from app.core.config import LOG_DIR

SENSITIVE = {"authorization", "token", "password", "api_key", "anthropic_api_key"}

def _redact(value):
    if isinstance(value, dict):
        return {k: ("<redacted>" if k.lower() in SENSITIVE else _redact(v)) for k, v in value.items()}
    if isinstance(value, list):
        return [_redact(v) for v in value]
    return value

def audit(event: str, payload: dict):
    request_id = payload.get("request_id") or str(uuid.uuid4())
    entry = {"ts": datetime.now(timezone.utc).isoformat(), "event": event, "request_id": request_id, "payload": _redact(payload)}
    path = Path(LOG_DIR) / "audit"
    path.mkdir(parents=True, exist_ok=True)
    with (path / "audit.jsonl").open("a", encoding="utf-8") as f:
        f.write(json.dumps(entry, ensure_ascii=False) + "\n")
    return request_id
